# Importing the required libraries

In [1]:
# Import required libraries
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.layers import Input


# Load the data

In [2]:
# Load the data
data = pd.read_csv('https://cocl.us/concrete_data')

data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

# Split the data into predictors and target

In [3]:
X = data.drop(columns=['Strength'])
y = data['Strength']

# Normalize the data

In [4]:
from sklearn.preprocessing import StandardScaler

# Normalize the data
scaler = StandardScaler()
X_normalized = scaler.fit_transform(X)

# Build a neural network model

In [5]:
#Build a baseline model 
def build_model():
    model = Sequential([
        Input(shape=(X.shape[1],)),  # Define the input shape using an Input layer
        Dense(10, activation='relu'),  # Hidden layer with 10 nodes and ReLU activation
        Dense(10, activation='relu'),  # Hidden layer with 10 nodes and ReLU activation
        Dense(10, activation='relu'),  # Hidden layer with 10 nodes and ReLU activation
        Dense(1)  # Output layer with 1 node (regression task)
    ])
    model.compile(optimizer=Adam(), loss='mean_squared_error')
    return model

# Test the model

In [6]:
# Train and evaluate the model
def train_and_evaluate():
    X_train, X_test, y_train, y_test = train_test_split(X_normalized, y, test_size=0.3, random_state=np.random.randint(10000))
    model = build_model()
    model.fit(X_train, y_train, epochs=50, batch_size=10, verbose=0)  # Set verbose to 0 to suppress output
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test, y_pred)
    return mse


# Train and evaluate the model

In [7]:
mse_list = []
for i in range(50):
    mse = train_and_evaluate()
    mse_list.append(mse)
    print(f'Mean Squared Error {i+1}: {mse}')
    

10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 1: 122.09585752510736
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
Mean Squared Error 2: 61.792936293297714
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 3: 72.28301631265815
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 4: 70.49243468799226
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 5: 124.5256833151369
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step 
Mean Squared Error 6: 104.81754400544484
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 7: 66.43563907234805
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 8: 88.34070397781291
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 9: 64.25788480733965
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 10: 70.13225305984825
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 11: 92.42826570814783
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 4ms/step 
Mean Squared Error 12: 79.50357835287042
10/10 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/

# Calculate the mean and standard deviation of the mean squared errors

In [8]:

# Calculate the mean and standard deviation of the mean squared errors
mse_list = np.array(mse_list)
mean = mse_list.mean()
std = mse_list.std()
print(f'Mean of mean squared errors: {mean}')
print(f'Standard deviation of mean squared errors: {std}')


Mean of mean squared errors: 83.7807964091929
Standard deviation of mean squared errors: 20.260058417799556


# How does the mean of the mean squared errors compare to that from Step A?

## In step B, the mean of the mean squared errors is 130.05 and the standard deviation is 11.77. In this step the mean of the mean squared errors is 83.78 and the standard deviation is 20.26. The mean of the mean squared errors is lower in this step compared to step B. The standard deviation is higher in this step compared to step B. The model in this step is more consistent in terms of performance compared to the model in step B. The model in this step is better than the model in step B.